### PR TITLE
Implement the Load barrier 

### PR DESCRIPTION
**Pull request by youngar**
- Remove the SATB write barrier
- Add the concurrent mark load barrier
- Add the concurrent compact load barrier
- Remove the ability to toggle the barriers

### CODE REVIEW

File: gc/include/omtalk/Barrier.h [hunk 1]
```diff
@@ -11,34 +13,59 @@ namespace omtalk::gc {
...
+        available = buffer.available();
+        result = copy(gcContext, address, to, available);
+      }
+
+      entry.set(to);
+      address = to;
+    } else {
+      // Object is being or has already been copied. entry.get() will wait for
+      // the object to be copied.
+      address = entry.get();
```
Comment (youngar) [inline gc/include/omtalk/Barrier.h:53]: If the thread doing the copying stalls out for an allocation buffer, this thread will never hit a gc sync point and the whole thing will deadlock.

File: gc/include/omtalk/Barrier.h [hunk 2]
```diff
@@ -11,34 +13,59 @@ namespace omtalk::gc {
...
+      GlobalCollectorContext<S> &gcContext = context.getCollectorContext();
+      auto &buffer = context.buffer();
+      auto *to = buffer.begin;
+      auto available = buffer.available();
+      auto result = copy(gcContext, ObjectProxy<S>(address), to, available);
+
+      if (!result) {
+        // get a new allocation buffer
+        auto copySize = getCopySize<S>(address);
+        context.refreshBuffer(copySize);
```
Comment (youngar) [inline gc/include/omtalk/Barrier.h:40]: I have to find a better way to get a region to copy to.

File: gc/example/include/example/Object.h [hunk 3]
```diff
@@ -348,6 +361,18 @@ struct gc::RootWalker<TestCollectorScheme> {
 // Test Allocator
 //===----------------------------------------------------------------------===//
 
+inline gc::Ref<TestObject> load(gc::Context<TestCollectorScheme> &cx,
+                                gc::Ref<TestStructObject> ref,
+                                std::size_t slot) {
+  TestObjectProxy objectProxy(ref);
+  TestSlotProxy slotProxy(ref->getSlot(slot));
+  Ref<void> result = omtalk::gc::load<S>(context, slotProxy);
```
Comment (rwy0717) [inline gc/example/include/example/Object.h:369]: what is gc::load<S>? does this need a memory order parameter?

File: gc/example/include/example/Object.h [hunk 4]
```diff
@@ -348,6 +361,18 @@ struct gc::RootWalker<TestCollectorScheme> {
...
+}
+
+inline gc::Ref<TestObject> store(gc::Contexted<TestCollectorScheme> &cx,
+                                 gc::Ref<TestStructObject> object,
+                                 std::size_t slot) {}
```
Comment (rwy0717) [inline gc/example/include/example/Object.h:374]: unimplemented and, missing a value argument? Does this need a memory order?

File: gc/example/include/example/Object.h [hunk 5]
```diff
@@ -348,6 +361,18 @@ struct gc::RootWalker<TestCollectorScheme> {
...
 
+inline gc::Ref<TestObject> load(gc::Context<TestCollectorScheme> &cx,
+                                gc::Ref<TestStructObject> ref,
+                                std::size_t slot) {
+  TestObjectProxy objectProxy(ref);
+  TestSlotProxy slotProxy(ref->getSlot(slot));
+  Ref<void> result = omtalk::gc::load<S>(context, slotProxy);
+}
+
+inline gc::Ref<TestObject> store(gc::Contexted<TestCollectorScheme> &cx,
```
Comment (rwy0717) [inline gc/example/include/example/Object.h:372]:
```suggestion
inline gc::Ref<TestObject> store(gc::Context<TestCollectorScheme> &cx,
```

File: gc/include/omtalk/Barrier.h [hunk 6]
```diff
@@ -11,34 +13,59 @@ namespace omtalk::gc {
 template <typename S, typename ObjectProxyT>
 void allocateBarrier(Context<S> &context, ObjectProxyT object) {
   // Newly allocated objects must be marked black and scanned by the GC.
-  if (context.writeBarrierEnabled()) {
-    mark<S>(context.getCollectorContext(), object);
-  }
+  mark<S>(context.getCollectorContext(), object);
```
Comment (rwy0717) [inline gc/include/omtalk/Barrier.h:16]: Can we turn off the allocate barrier during evacuate?

File: gc/include/omtalk/Barrier.h (hunk 2 above)
Comment (youngar) [reply gc/include/omtalk/Barrier.h:42]: Refresh buffer now does not cause a GC, and will allocate more memory.

**Pull request merged**